### ISSUE TITLE
Migrating application to Cloud

### ISSUE THREAD

**Issue by AvJain** [borisliao/mbox#36]
Make the necessary changes required to move the application to Heroku (alternative AWS) 

Steps: 
1. Add logging to application to measure cpu and memory usage 
2. Based on metrics decide whether we want to migrate app to AWS or Heroku 

Potential things to consider
AWS allows continues delivery through code pipeline
We can use the AWS free tier  (1 year and build the pipeline)

**Comment by borisliao** [comment]
Although I like the idea of AWS, Heroku dyno running 24/7 seams to be the easiest option for now for a pipeline production standpoint.

[AWS EC2](https://aws.amazon.com/ec2/) seems like what we would use if we use AWS however there might be other options since the bot is not responding to requests most of the time. Most on the fly cloud computing does not work since discord requires a live websocket connection.

**Comment by Exsects** [comment]
I agree with most of your thoughts that running on Heroku, we can set a worker dyno to run 24/7. If we need to scale, we can always switch over to AWS.

For a year, AWS provides us with a free tier option in which provides us with 750 hours for EC2 as well a t2.micro (1GB of RAM). 750 hours is enough to last us a month assuming we are not spinning up another EC2 instance, otherwise the 750 hours would be split 50/50. 

I'm not sure if I understood your last point where you mentioned "Most on the fly cloud computing does not work since discord requires a live websocket connection." As long as we don't terminate the EC2 instance, we are essentially keeping the server on 24/7 preserving the live web socket connection.

**Comment by borisliao** [comment]
> Most on the fly cloud computing does not work since discord requires a live websocket connection.

Options like Google Cloud function, Amazon Lambda, or even Heroku dynos going to sleep automatically would not work well for this bot since the Discord API requires a live websocket connection, and not providing a "heartbeat" ping to Discord would result in our bot going offline and the program not receiving events (message events) from Discord.

This is most evident when you run the program through a debugger and activate a breakpoint. If you do not let the bot continue execution, it will warn you that the shard heartbeat was blocked for x seconds. Let it be blocked for a while (about 40 seconds to 60 seconds) and discord will kick you off, forcing your bot to log in again and start a new websocket connection. 
[Heartbeat Policy on the Discord Developer Portal](https://discord.com/developers/docs/topics/gateway#heartbeating)
![Shard Blocked](https://user-images.githubusercontent.com/16157348/117090497-76e4ca00-ad26-11eb-8041-c72fce394443.png)

TLDR: All this is to say that until interactions support becomes well built out in the discord.py library where we could just run a lambda function [based on a webhook](https://discord.com/developers/docs/interactions/slash-commands#receiving-an-interaction) we will need to run the bot with a 24/7 dyno or a AWS EC2 instance. Luckily that is not a problem, but I just don't know if there is any more economical option (I don't think so).

**Comment by Exsects** [comment]
If we were strictly doing slash commands and they're not actions that require us to preserve our connection to the Discord, it could have been more economically. A simple example would be /ping, where it just returns one value and doesn't rely on the chat history. In that case, we could have done API Gateway + Lambda. 

In our case since we need to maintain a persistent connection to the voice channel for Music Box to work, using AWS Lambda or Cloud Functions won't fit our requirements. So a Heroku worker dyno or EC2 would be our best bet. 

For Heroku, there is a configuration where if you set it to web dyno, it automatically shuts off after inactivity. However, if you turn off web dyno and only have worker dyno turn on, you will have it on 24/7.
